doc/advanced/content/executor_design.rst

```diff
@@ -0,0 +1,417 @@
+absent from sequential implementations.
+
+3. **Inextensible Design**
+The current design doesn't support using new facilities like thread pools, multi-gpu support or nesting
+facilities. To add support for these facilities completely new implementations will have to be written
+for every algorithm.
+Example: It isn't possible to run vectorized code (SIMD) while running a parallel implementation which
+uses OpenMP.
+
+4. **Code Duplication**
+Even if different facilities might require slightly different implementations, a lot of the code can
+be shared. Having different implementations just leads to a majority of the code base being duplicated
+and only some of the code gets modified inorder to adapt to the interface provided by the facility.
+Example: Most of OpenMP code is quite similar to the sequential implementation with only some additions.
+So having separate classed for OpenMP implementations is quite redundant.
+
+5. **Maintenance Overhead**
+Maintaining several implementations of the same algorithm is a labour and time intensive task.
+Porting changes from one implementation to another and propagating bug fixes to all the implementations
+is time consuming. Lack of time to propagate the changes will lead to divergent implementations as mentioned
+above.
+
+Design Considerations
+=================================
+
+The executor design proposal for C++, aims to build a generalized and extensible framework surrounding
+executors. This is necessary since it needs to support a wide rage of uses cases so as to attempt to cater
+everyone's needs, in order for it to be accepted into the standard library.
+
+In PCL, the need for executors is limited to certain features and there is no need for the entire feature set
+as mentioned in the proposal. Concepts such asynchronous operations and task based parallelism are not present
+or needed in PCL at the moment so creating a design which incorporates all those features would be unnecessary.
+The main use cases in PCL are:
+
+1. Provide a uniform API for executing existing algorithms on different facilities giving users the freedom
+   to switch between facilities with ease.
+
+2. Reducing code duplication and trying to avoid completely different implementations of the same algorithm
+
+3. Provide a simple and easy to use mechanism to customize the execution context which users can also access
+
+4. Expose some of the underlying features offered by the the various facilities in a standardized manner
+
+5. Provide a mechanism to automatically choose the best facility to run an algorithm in case the user does
+   not explicitly specify which facility to use
+
+6. Be extensible enough to allows users to specify their own executors or customize the ones provided by PCL
+7. Ensure as little overhead as possible, if possible make everything compile time. This is also ensure all
+   errors are caught at compile time, providing guarantee at compile time that an executor would work.
```

Comment:
too much repetition of the expression compile time

Reply:
Reworded